documentation/multstep-form.md

```diff
@@ -0,0 +1,243 @@
+# Multistep form
+
+Multistep form is a mini-library aiming at reducing the complexity of long user journeys consisting of multiple steps.
+
+While many libraries exist for that reason already, they are all quite limited and require a number of hacks in order to, for example, create dynamic step orders. This usually leads to really difficult to read and maintain source code as the smallest change has unexpected consequences.
+
+## How is it different
+
+This library is based on dynamic step ordering and, hopefully, handles a number of common but yet interesting edge-cases, like user resubmitting one of the steps making other steps invalid.
+
+The library is still work in progress!
+
+## Multstep::Form
+
+To use multistep, you need to create a new class and include `Multistep::Form` module. Then simply define steps as isolated form objects:
+
+```ruby
+require 'multistep/form'
+
+class MyMultistepForm
+  include Multistep::Form
+
+  attribute :company_id
+  
+  step :name do
+    attribute :first_name
+    attribute :last_name
+
+    validates :first_name, :last_name, presence: true
+  end
+
+  step :contact do
+    attribute :email
+    attribute :lists, array: true
+
+    validates :email, presence: true
+  end
+  
+  step :another_step do
+    attribute :some_attribute
+  end
+end
+```
+
+### Attributes
+Attributes can be defined both within steps and within the Multistep object. Attributes represents the state of the form and all steps' attributes are automatically accessible on the object itself:
```

Comment:
Assume you mean the multistep form object here?

Reply:
Clarified

Reply:
Clarified